ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java

```diff
@@ -0,0 +1,777 @@
+  private final AtomicBoolean operationCallbackInvoked = new AtomicBoolean(false);
+  // The first chunk may be a metadata chunk if the blob is composite, or the only data chunk if the blob is simple.
+  private final FirstGetChunk firstChunk;
+  // Associated with all data chunks in the case of composite blobs. Only a fixed number of these are initialized.
+  // Each of these is initialized with the information required to fetch a data chunk and is responsible for
+  // retrieving and adding it to the list of chunk buffers. Once complete, they are reused to fetch subsequent data
+  // chunks.
+  private GetChunk[] dataChunks;
+  // the factory to use to deserialize keys in a metadata chunk.
+  private final BlobIdFactory blobIdFactory;
+  // the total number of data chunks associated with this blob.
+  private int numChunksTotal;
+  // the total number of data chunks retrieved so far (and may or may not have been written out yet).
+  private int numChunksRetrieved;
+  // a list iterator to the chunk ids that need to be fetched for this operation, if this is a composite blob.
+  private ListIterator<StoreKey> chunkIdIterator;
+  // chunk index to retrieved chunk buffer mapping.
+  private Map<Integer, ByteBuffer> chunkIndexToBuffer;
+  // To find the GetChunk to hand over the response quickly.
+  private final Map<Integer, GetChunk> correlationIdToGetChunk = new HashMap<>();
+
+  // ReadableStreamChannel implementation related variables:
+
+  // whether this ReadableStreamChannel is open.
+  private boolean isOpen = true;
+  // whether readInto() has been called yet by the caller on this ReadableStreamChannel.
+  private volatile boolean readCalled = false;
+  // The channel to write chunks of the blob into. This will be initialized when the caller calls the readInto().
+  private AsyncWritableChannel asyncWritableChannel;
+  // the callback to call when all the chunks are successfully written out into the asyncWritableChannel.
+  private Callback<Long> readIntoCallback;
+  // the future to mark as done when all the chunks are successfully written out into the asyncWritableChannel.
+  private FutureResult<Long> readIntoFuture;
+  // the number of bytes written out to the asyncWritableChannel. This would be the size of the blob eventually.
+  private Long bytesWritten = 0L;
+  // the number of chunks that have been written out to the asyncWritableChannel.
+  private final AtomicInteger numChunksWrittenOut = new AtomicInteger(0);
+  // the index of the next chunk that is to be written out to the asyncWritableChannel.
+  private int indexOfNextChunkToWriteOut = 0;
+  // whether this object has called the readIntoCallback yet.
+  private AtomicBoolean readIntoCallbackCalled = new AtomicBoolean(false);
+  // the callback that is passed into the asyncWritableChannel write() operation.
+  Callback<Long> chunkAsyncWriteCallback = new Callback<Long>() {
+    @Override
+    public void onCompletion(Long result, Exception exception) {
```

Comment:
It seemed logical to do that at first, but then it didn't seem like the right thing to do. Note that this method gets called in the context of another thread. We would then have to worry about making the methods that write out thread safe. It didn't seem worth it, as at the most there would be `MAX_IN_MEM_CHUNKS - 1` or so chunks ready to be written out. And in the current scheme, all of those will get written out in the very next poll.